File: static/js/account_settings.js
/**
 * Account Settings JavaScript
 * Handles all interactions for the account settings page
 */

class AccountSettings {
    constructor() {
        this.init();
    }

    init() {
        // Load data when page loads
        this.loadProfileData();
        this.loadPaymentMethods();
        this.loadSubscriptionDetails();
        this.loadNotificationPreferences();
        
        // Set up event listeners
        this.setupEventListeners();
    }

    setupEventListeners() {
        // Tab change listeners
        document.querySelectorAll('[data-bs-toggle="pill"]').forEach(tab => {
            tab.addEventListener('shown.bs.tab', (e) => {
                const targetTab = e.target.getAttribute('data-bs-target');
                this.onTabChange(targetTab);
            });
        });
    }

    onTabChange(targetTab) {
        // Load data specific to the active tab
        switch(targetTab) {
            case '#payments':
                this.loadPaymentMethods();
                break;
            case '#subscription':
                this.loadSubscriptionDetails();
                break;
            case '#notifications':
                this.loadNotificationPreferences();
                break;
        }
    }

    async loadProfileData() {
        try {
            const response = await fetch('/api/account/profile');
            const data = await response.json();
            
            if (data.error) {
                this.showError('Failed to load profile data');
                return;
            }

            // Populate profile form
            const userInfo = data.user_info || {};
            const accountInfo = data.account_info || {};

            document.getElementById('firstName').value = userInfo.first_name || '';
            document.getElementById('lastName').value = userInfo.last_name || '';
            document.getElementById('email').value = userInfo.email || '';
            
            if (userInfo.profile_image_url) {
                const img = document.getElementById('profileImage');
                img.src = userInfo.profile_image_url;
                img.style.display = 'block';
            }

            // Update statistics
            document.getElementById('totalTrades').textContent = accountInfo.total_trades || '0';
            document.getElementById('accountBalance').textContent = '$' + (accountInfo.balance || '0.00');
            document.getElementById('memberSince').textContent = accountInfo.member_since || 'Recently';
            document.getElementById('lastLogin').textContent = userInfo.last_login || 'Unknown';

        } catch (error) {
            console.error('Error loading profile:', error);
            this.showError('Failed to load profile data');
        }
    }

    async loadPaymentMethods() {
        try {
            const response = await fetch('/api/account/payment-methods');
            const data = await response.json();
            
            if (!data.success) {
                this.showError('Failed to load payment methods');
                return;
            }

            this.renderPaymentMethods(data.payment_methods);

        } catch (error) {
            console.error('Error loading payment methods:', error);
            this.showError('Failed to load payment methods');
        }
    }

    renderPaymentMethods(paymentMethods) {
        const container = document.getElementById('paymentMethodsList');
        const cards = paymentMethods.cards || [];
        
        if (cards.length === 0) {
            container.innerHTML = `
                <div class="text-center py-4">
                    <i class="fas fa-credit-card fa-3x text-muted mb-3"></i>
                    <h5 class="text-muted">No payment methods added</h5>
                    <p class="text-muted">Add a payment method to start trading</p>
                </div>
            `;
            return;
        }

        container.innerHTML = cards.map(card => `
            <div class="payment-method-card ${card.is_default ? 'default' : ''}">
                <div class="d-flex justify-content-between align-items-center">
                    <div class="d-flex align-items-center">
                        <i class="fab fa-cc-${card.brand} fa-2x me-3 text-primary"></i>
                        <div>
                            <h6 class="mb-1">•••• •••• •••• ${card.last4}</h6>
                            <small class="text-muted">Expires ${card.exp_month}/${card.exp_year}</small>
                            ${card.is_default ? '<br><small class="text-success"><i class="fas fa-check"></i> Default</small>' : ''}
                        </div>
                    </div>
                    <div class="btn-group">
                        ${!card.is_default ? '<button class="btn btn-sm btn-outline-primary" onclick="setDefaultPaymentMethod(\'' + card.id + '\')">Set Default</button>' : ''}
                        <button class="btn btn-sm btn-outline-danger" onclick="removePaymentMethod('${card.id}')">
                            <i class="fas fa-trash"></i>
                        </button>
                    </div>
                </div>
                ${card.billing_address ? `
                    <div class="mt-2">
                        <small class="text-muted">
                            ${card.billing_address.line1}<br>
                            ${card.billing_address.city}, ${card.billing_address.state} ${card.billing_address.postal_code}
                        </small>
                    </div>
                ` : ''}
            </div>
        `).join('');
    }

    async loadSubscriptionDetails() {
        try {
            const response = await fetch('/api/account/subscription');
            const data = await response.json();
            
            if (!data.success) {
                this.showError('Failed to load subscription details');
                return;
            }

            this.renderSubscriptionDetails(data.subscription);

        } catch (error) {
            console.error('Error loading subscription:', error);
            this.showError('Failed to load subscription details');
        }
    }

    renderSubscriptionDetails(subscription) {
        const currentPlan = subscription.current_plan || {};
        const usageStats = subscription.usage_stats || {};
        const availablePlans = subscription.available_plans || [];

        // Render current plan
        document.getElementById('currentPlan').innerHTML = `
            <div class="plan-badge">Current</div>
            <h5>${currentPlan.name} Plan</h5>
            <h3 class="text-primary">$${currentPlan.price}/month</h3>
            <p class="text-muted">Next billing: ${currentPlan.next_billing_date}</p>
            <ul class="list-unstyled mt-3">
                ${(currentPlan.features || []).map(feature => `<li><i class="fas fa-check text-success me-2"></i>${feature}</li>`).join('')}
            </ul>
        `;

        // Update usage statistics
        const apiUsagePercent = Math.min(100, (usageStats.api_calls_used / usageStats.api_calls_limit) * 100);
        document.getElementById('apiUsageBar').style.width = `${apiUsagePercent}%`;
        document.getElementById('apiUsageText').textContent = `${usageStats.api_calls_used} / ${usageStats.api_calls_limit} calls used`;
        
        const tradesUsagePercent = Math.min(100, (usageStats.trades_this_month / 100) * 100); // Assuming 100 as max
        document.getElementById('tradesUsageBar').style.width = `${tradesUsagePercent}%`;
        document.getElementById('tradesUsageText').textContent = `${usageStats.trades_this_month} trades this month`;

        // Render available plans
        document.getElementById('availablePlans').innerHTML = availablePlans.map(plan => `
            <div class="subscription-plan ${plan.name === currentPlan.name ? 'current' : ''} ${plan.recommended ? 'recommended' : ''}">
                ${plan.recommended ? '<div class="plan-badge recommended">Recommended</div>' : ''}
                <div class="d-flex justify-content-between align-items-start">
                    <div>
                        <h5>${plan.name} Plan</h5>
                        <h4 class="text-primary mb-3">$${plan.price}/month</h4>
                        <ul class="list-unstyled">
                            ${plan.features.map(feature => `<li><i class="fas fa-check text-success me-2"></i>${feature}</li>`).join('')}
                        </ul>
                    </div>
                    <div>
                        ${plan.name !== currentPlan.name ? `
                            <button class="btn btn-primary" onclick="changeSubscription('${plan.name}')">
                                ${plan.price > currentPlan.price ? 'Upgrade' : 'Downgrade'}
                            </button>
                        ` : '<span class="badge bg-success">Current Plan</span>'}
                    </div>
                </div>
            </div>
        `).join('');
    }

    async loadNotificationPreferences() {
        try {
            const response = await fetch('/api/account/notifications');
            const data = await response.json();
            
            if (!data.success) {
                this.showError('Failed to load notification preferences');
                return;
            }

            this.applyNotificationPreferences(data.preferences);

        } catch (error) {
            console.error('Error loading notifications:', error);
            this.showError('Failed to load notification preferences');
        }
    }

    applyNotificationPreferences(preferences) {
        const emailPrefs = preferences.email_notifications || {};
        const pushPrefs = preferences.push_notifications || {};
        const alertFreq = preferences.alert_frequency || {};

        // Set email preferences
        Object.keys(emailPrefs).forEach(key => {
            const element = document.getElementById(`email${key.charAt(0).toUpperCase() + key.slice(1).replace(/_(.)/g, (_, letter) => letter.toUpperCase())}`);
            if (element) element.checked = emailPrefs[key];
        });

        // Set push preferences
        Object.keys(pushPrefs).forEach(key => {
            const element = document.getElementById(`push${key.charAt(0).toUpperCase() + key.slice(1).replace(/_(.)/g, (_, letter) => letter.toUpperCase())}`);
            if (element) element.checked = pushPrefs[key];
        });

        // Set alert frequency
        Object.keys(alertFreq).forEach(key => {
            const element = document.getElementById(key);
            if (element && alertFreq[key]) element.checked = true;
        });
    }

    showError(message) {
        // Create and show error alert
        const alert = document.createElement('div');
        alert.className = 'alert alert-danger alert-dismissible fade show';
        alert.innerHTML = `
            ${message}
            <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
        `;
        
        const container = document.querySelector('.tab-content');
        container.insertBefore(alert, container.firstChild);
    }

    showSuccess(message) {
        // Create and show success alert
        const alert = document.createElement('div');
        alert.className = 'alert alert-success alert-dismissible fade show';
        alert.innerHTML = `
            ${message}
            <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
        `;
        
        const container = document.querySelector('.tab-content');
        container.insertBefore(alert, container.firstChild);
    }

    showLoading(show = true) {
        const spinner = document.getElementById('loadingSpinner');
        spinner.style.display = show ? 'block' : 'none';
    }
}

// Global functions for button interactions
async function updateProfile() {
    const accountSettings = window.accountSettings;
    
    const profileData = {
        first_name: document.getElementById('firstName').value,
        last_name: document.getElementById('lastName').value,
        email: document.getElementById('email').value
    };

    try {
        accountSettings.showLoading(true);
        const response = await fetch('/api/account/profile', {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(profileData)
        });

        const result = await response.json();
        
        if (result.success) {
            accountSettings.showSuccess('Profile updated successfully!');
        } else {
            accountSettings.showError(result.error || 'Failed to update profile');
        }
    } catch (error) {
        console.error('Error updating profile:', error);
        accountSettings.showError('Failed to update profile');
    } finally {
        accountSettings.showLoading(false);
    }
}

async function addPaymentMethod() {
    // In a real implementation, this would integrate with Stripe Elements
    const accountSettings = window.accountSettings;
    accountSettings.showSuccess('Payment method integration coming soon!');
}

async function removePaymentMethod(paymentMethodId) {
    if (!confirm('Are you sure you want to remove this payment method?')) {
        return;
    }

    const accountSettings = window.accountSettings;
    
    try {
        const response = await fetch(`/api/account/payment-methods/${paymentMethodId}`, {
            method: 'DELETE'
        });

        const result = await response.json();
        
        if (result.success) {
            accountSettings.showSuccess('Payment method removed successfully!');
            accountSettings.loadPaymentMethods(); // Reload the list
        } else {
            accountSettings.showError(result.error || 'Failed to remove payment method');
        }
    } catch (error) {
        console.error('Error removing payment method:', error);
        accountSettings.showError('Failed to remove payment method');
    }
}

async function changeSubscription(planName) {
    if (!confirm(`Are you sure you want to change to the ${planName} plan?`)) {
        return;
    }

    const accountSettings = window.accountSettings;
    
    try {
        accountSettings.showLoading(true);
        const response = await fetch('/api/account/subscription', {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({ plan: planName })
        });

        const result = await response.json();
        
        if (result.success) {
            accountSettings.showSuccess(`Successfully changed to ${planName} plan!`);
            accountSettings.loadSubscriptionDetails(); // Reload subscription details
        } else {
            accountSettings.showError(result.error || 'Failed to change subscription');
        }
    } catch (error) {
        console.error('Error changing subscription:', error);
        accountSettings.showError('Failed to change subscription');
    } finally {
        accountSettings.showLoading(false);
    }
}

async function updateSecuritySettings() {
    const accountSettings = window.accountSettings;
    
    const securityData = {
        two_factor_enabled: document.getElementById('twoFactorAuth').checked,
        login_notifications: document.getElementById('loginNotifications').checked,
        trading_alerts: document.getElementById('tradingAlerts').checked
    };

    try {
        accountSettings.showLoading(true);
        const response = await fetch('/api/account/security', {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(securityData)
        });

        const result = await response.json();
        
        if (result.success) {
            accountSettings.showSuccess('Security settings updated successfully!');
        } else {
            accountSettings.showError(result.error || 'Failed to update security settings');
        }
    } catch (error) {
        console.error('Error updating security settings:', error);
        accountSettings.showError('Failed to update security settings');
    } finally {
        accountSettings.showLoading(false);
    }
}

async function updateNotificationPreferences() {
    const accountSettings = window.accountSettings;
    
    const preferences = {
        email_notifications: {
            trade_confirmations: document.getElementById('emailTradeConfirmations').checked,
            market_alerts: document.getElementById('emailMarketAlerts').checked,
            account_updates: document.getElementById('emailAccountUpdates').checked,
            promotional_emails: document.getElementById('emailPromotional').checked,
            weekly_summary: document.getElementById('emailWeeklySummary').checked
        },
        push_notifications: {
            price_alerts: document.getElementById('pushPriceAlerts').checked,
            trade_executions: document.getElementById('pushTradeExecutions').checked,
            market_news: document.getElementById('pushMarketNews').checked,
            account_security: document.getElementById('pushAccountSecurity').checked
        },
        alert_frequency: {
            immediate: document.getElementById('immediate').checked,
            hourly_digest: document.getElementById('hourlyDigest').checked,
            daily_digest: document.getElementById('dailyDigest').checked,
            weekly_digest: document.getElementById('weeklyDigest').checked
        }
    };

    try {
        accountSettings.showLoading(true);
        const response = await fetch('/api/account/notifications', {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(preferences)
        });

        const result = await response.json();
        
        if (result.success) {
            accountSettings.showSuccess('Notification preferences updated successfully!');
        } else {
            accountSettings.showError(result.error || 'Failed to update preferences');
        }
    } catch (error) {
        console.error('Error updating notifications:', error);
        accountSettings.showError('Failed to update notification preferences');
    } finally {
        accountSettings.showLoading(false);
    }
}

async function changePassword() {
    const currentPassword = document.getElementById('currentPassword').value;
    const newPassword = document.getElementById('newPassword').value;
    
    if (!currentPassword || !newPassword) {
        window.accountSettings.showError('Please fill in both password fields');
        return;
    }

    if (newPassword.length < 8) {
        window.accountSettings.showError('New password must be at least 8 characters long');
        return;
    }

    // In a real implementation, this would be handled securely
    window.accountSettings.showSuccess('Password change functionality coming soon!');
}

async function deleteAccount() {
    const email = prompt('To confirm account deletion, please enter your email address:');
    
    if (!email) {
        return;
    }

    if (!confirm('This action cannot be undone. Are you absolutely sure you want to delete your account?')) {
        return;
    }

    const accountSettings = window.accountSettings;
    
    try {
        accountSettings.showLoading(true);
        const response = await fetch('/api/account/delete', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                email: email,
                confirmed: true
            })
        });

        const result = await response.json();
        
        if (result.success) {
            alert('Account deletion initiated. You will receive confirmation within 24 hours.');
            // In a real implementation, redirect to logout
        } else {
            accountSettings.showError(result.error || 'Failed to delete account');
        }
    } catch (error) {
        console.error('Error deleting account:', error);
        accountSettings.showError('Failed to delete account');
    } finally {
        accountSettings.showLoading(false);
    }
}

// Initialize when DOM is loaded
document.addEventListener('DOMContentLoaded', function() {
    window.accountSettings = new AccountSettings();
});

File: account_settings.py
"""
Account Settings Management System
Provides comprehensive user account management capabilities including payment methods,
profile settings, security options, and subscription management.
"""

import os
import stripe
from datetime import datetime, timedelta
from typing import Dict, List, Any, Optional
from models import db, User, UserAccount, Transaction
from flask import current_app
import logging

# Configure Stripe
stripe.api_key = os.environ.get('STRIPE_SECRET_KEY')

class AccountSettingsManager:
    """Comprehensive account settings management system"""
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        
    def get_user_profile(self, user_id: str) -> Dict[str, Any]:
        """Get complete user profile information"""
        try:
            user = User.query.filter_by(id=user_id).first()
            if not user:
                return {'error': 'User not found'}
            
            # Get account information
            account = UserAccount.query.filter_by(user_id=user_id).first()
            
            # Get subscription information
            subscription_info = self._get_subscription_info(user_id)
            
            # Get recent transactions (handle timestamp field compatibility)
            try:
                recent_transactions = Transaction.query.filter_by(user_id=user_id).order_by(Transaction.created_at.desc()).limit(5).all()
            except Exception:
                # Fallback if Transaction table/field doesn't exist
                recent_transactions = []
            
            profile_data = {
                'user_info': {
                    'id': user.id,
                    'email': user.email,
                    'first_name': getattr(user, 'first_name', user.username),
                    'last_name': getattr(user, 'last_name', ''),
                    'profile_image_url': getattr(user, 'profile_image_url', '/static/images/default-avatar.png'),
                    'created_at': user.created_at.strftime('%Y-%m-%d') if hasattr(user, 'created_at') else 'Unknown',
                    'last_login': datetime.now().strftime('%Y-%m-%d %H:%M')
                },
                'account_info': {
                    'balance': float(account.balance) if account else 0.0,
                    'total_deposits': float(account.total_deposited) if account else 0.0,
                    'total_trades': len(recent_transactions),
                    'account_status': 'Active',
                    'account_type': subscription_info.get('tier', 'Free'),
                    'member_since': user.created_at.strftime('%B %Y') if hasattr(user, 'created_at') else 'Recently'
                },
                'subscription': subscription_info,
                'recent_activity': [
                    {
                        'type': getattr(t, 'transaction_type', 'trade'),
                        'amount': float(getattr(t, 'amount', 0)),
                        'description': getattr(t, 'description', f'Trade activity'),
                        'timestamp': getattr(t, 'created_at', datetime.now()).strftime('%Y-%m-%d %H:%M')
                    } for t in recent_transactions
                ],
                'security_settings': {
                    'two_factor_enabled': False,  # To be implemented
                    'login_notifications': True,
                    'trading_alerts': True,
                    'email_notifications': True
                }
            }
            
            return profile_data
            
        except Exception as e:
            self.logger.error(f"Error getting user profile: {e}")
            return {'error': 'Failed to load profile'}
    
    def update_profile(self, user_id: str, profile_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update user profile information"""
        try:
            user = User.query.filter_by(id=user_id).first()
            if not user:
                return {'success': False, 'error': 'User not found'}
            
            # Update allowed fields
            if 'first_name' in profile_data:
                user.first_name = profile_data['first_name']
            if 'last_name' in profile_data:
                user.last_name = profile_data['last_name']
            if 'email' in profile_data:
                # Validate email format
                if '@' in profile_data['email']:
                    user.email = profile_data['email']
                else:
                    return {'success': False, 'error': 'Invalid email format'}
            
            db.session.commit()
            
            return {
                'success': True,
                'message': 'Profile updated successfully',
                'updated_fields': list(profile_data.keys())
            }
            
        except Exception as e:
            db.session.rollback()
            self.logger.error(f"Error updating profile: {e}")
            return {'success': False, 'error': 'Failed to update profile'}
    
    def get_payment_methods(self, user_id: str) -> Dict[str, Any]:
        """Get user's saved payment methods"""
        try:
            # In a real implementation, this would fetch from Stripe
            # For now, return structured data for the interface
            
            payment_methods = {
                'cards': [
                    {
                        'id': 'pm_card_1',
                        'type': 'card',
                        'brand': 'visa',
                        'last4': '4242',
                        'exp_month': 12,
                        'exp_year': 2026,
                        'is_default': True,
                        'billing_address': {
                            'line1': '123 Main St',
                            'city': 'San Francisco',
                            'state': 'CA',
                            'postal_code': '94105',
                            'country': 'US'
                        }
                    }
                ],
                'bank_accounts': [],  # To be implemented
                'default_payment_method': 'pm_card_1'
            }
            
            return {
                'success': True,
                'payment_methods': payment_methods,
                'stripe_publishable_key': os.environ.get('STRIPE_PUBLISHABLE_KEY', 'pk_test_...')
            }
            
        except Exception as e:
            self.logger.error(f"Error getting payment methods: {e}")
            return {'success': False, 'error': 'Failed to load payment methods'}
    
    def add_payment_method(self, user_id: str, payment_data: Dict[str, Any]) -> Dict[str, Any]:
        """Add new payment method"""
        try:
            # In a real implementation, this would create a payment method in Stripe
            
            return {
                'success': True,
                'message': 'Payment method added successfully',
                'payment_method_id': f'pm_new_{user_id}'
            }
            
        except Exception as e:
            self.logger.error(f"Error adding payment method: {e}")
            return {'success': False, 'error': 'Failed to add payment method'}
    
    def remove_payment_method(self, user_id: str, payment_method_id: str) -> Dict[str, Any]:
        """Remove payment method"""
        try:
            # In a real implementation, this would detach the payment method in Stripe
            
            return {
                'success': True,
                'message': 'Payment method removed successfully'
            }
            
        except Exception as e:
            self.logger.error(f"Error removing payment method: {e}")
            return {'success': False, 'error': 'Failed to remove payment method'}
    
    def update_security_settings(self, user_id: str, security_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update security settings"""
        try:
            # Store security preferences (in a real app, this would be in the database)
            user = User.query.filter_by(id=user_id).first()
            if not user:
                return {'success': False, 'error': 'User not found'}
            
            # Update security settings
            settings_updated = []
            
            if 'two_factor_enabled' in security_data:
                # Implement 2FA logic here
                settings_updated.append('Two-factor authentication')
            
            if 'login_notifications' in security_data:
                settings_updated.append('Login notifications')
            
            if 'trading_alerts' in security_data:
                settings_updated.append('Trading alerts')
            
            if 'email_notifications' in security_data:
                settings_updated.append('Email notifications')
            
            return {
                'success': True,
                'message': f'Updated: {", ".join(settings_updated)}',
                'updated_settings': settings_updated
            }
            
        except Exception as e:
            self.logger.error(f"Error updating security settings: {e}")
            return {'success': False, 'error': 'Failed to update security settings'}
    
    def get_subscription_details(self, user_id: str) -> Dict[str, Any]:
        """Get detailed subscription information"""
        try:
            subscription_info = self._get_subscription_info(user_id)
            
            # Calculate subscription metrics
            current_tier = subscription_info.get('tier', 'Free')
            
            subscription_details = {
                'current_plan': {
                    'name': current_tier,
                    'price': subscription_info.get('monthly_cost', 0),
                    'billing_cycle': 'Monthly',
                    'next_billing_date': (datetime.now() + timedelta(days=30)).strftime('%Y-%m-%d'),
                    'features': self._get_tier_features(current_tier)
                },
                'usage_stats': {
                    'api_calls_used': subscription_info.get('api_calls_used', 0),
                    'api_calls_limit': subscription_info.get('api_calls_limit', 1000),
                    'trades_this_month': 15,  # Example data
                    'alerts_active': 8
                },
                'available_plans': [
                    {
                        'name': 'Free',
                        'price': 0,
                        'features': self._get_tier_features('Free'),
                        'recommended': False
                    },
                    {
                        'name': 'Pro',
                        'price': 19.99,
                        'features': self._get_tier_features('Pro'),
                        'recommended': current_tier == 'Free'
                    },
                    {
                        'name': 'Elite',
                        'price': 39.99,
                        'features': self._get_tier_features('Elite'),
                        'recommended': current_tier in ['Free', 'Pro']
                    },
                    {
                        'name': 'Institutional',
                        'price': 199.99,
                        'features': self._get_tier_features('Institutional'),
                        'recommended': False
                    }
                ]
            }
            
            return {'success': True, 'subscription': subscription_details}
            
        except Exception as e:
            self.logger.error(f"Error getting subscription details: {e}")
            return {'success': False, 'error': 'Failed to load subscription details'}
    
    def change_subscription(self, user_id: str, new_plan: str) -> Dict[str, Any]:
        """Change user subscription plan"""
        try:
            # In a real implementation, this would update the subscription in Stripe
            
            plan_prices = {
                'Free': 0,
                'Pro': 19.99,
                'Elite': 39.99,
                'Institutional': 199.99
            }
            
            if new_plan not in plan_prices:
                return {'success': False, 'error': 'Invalid plan selected'}
            
            return {
                'success': True,
                'message': f'Successfully upgraded to {new_plan} plan',
                'new_plan': new_plan,
                'new_price': plan_prices[new_plan],
                'effective_date': datetime.now().strftime('%Y-%m-%d')
            }
            
        except Exception as e:
            self.logger.error(f"Error changing subscription: {e}")
            return {'success': False, 'error': 'Failed to change subscription'}
    
    def get_notification_preferences(self, user_id: str) -> Dict[str, Any]:
        """Get user notification preferences"""
        try:
            # Default notification preferences
            preferences = {
                'email_notifications': {
                    'trade_confirmations': True,
                    'market_alerts': True,
                    'account_updates': True,
                    'promotional_emails': False,
                    'weekly_summary': True
                },
                'push_notifications': {
                    'price_alerts': True,
                    'trade_executions': True,
                    'market_news': False,
                    'account_security': True
                },
                'alert_frequency': {
                    'immediate': True,
                    'hourly_digest': False,
                    'daily_digest': True,
                    'weekly_digest': True
                }
            }
            
            return {'success': True, 'preferences': preferences}
            
        except Exception as e:
            self.logger.error(f"Error getting notification preferences: {e}")
            return {'success': False, 'error': 'Failed to load preferences'}
    
    def update_notification_preferences(self, user_id: str, preferences: Dict[str, Any]) -> Dict[str, Any]:
        """Update notification preferences"""
        try:
            # In a real implementation, store these in the database
            
            updated_categories = []
            if 'email_notifications' in preferences:
                updated_categories.append('Email notifications')
            if 'push_notifications' in preferences:
                updated_categories.append('Push notifications')
            if 'alert_frequency' in preferences:
                updated_categories.append('Alert frequency')
            
            return {
                'success': True,
                'message': f'Updated: {", ".join(updated_categories)}',
                'updated_preferences': updated_categories
            }
            
        except Exception as e:
            self.logger.error(f"Error updating notification preferences: {e}")
            return {'success': False, 'error': 'Failed to update preferences'}
    
    def delete_account(self, user_id: str, confirmation_data: Dict[str, Any]) -> Dict[str, Any]:
        """Delete user account with proper confirmation"""
        try:
            user = User.query.filter_by(id=user_id).first()
            if not user:
                return {'success': False, 'error': 'User not found'}
            
            # Verify confirmation
            if not confirmation_data.get('confirmed'):
                return {'success': False, 'error': 'Account deletion must be confirmed'}
            
            if confirmation_data.get('email') != user.email:
                return {'success': False, 'error': 'Email confirmation does not match'}
            
            # In a real implementation:
            # 1. Cancel all subscriptions
            # 2. Close all positions
            # 3. Transfer remaining balance
            # 4. Anonymize or delete data according to regulations
            
            return {
                'success': True,
                'message': 'Account deletion initiated. You will receive confirmation within 24 hours.',
                'deletion_id': f'del_{user_id}_{int(datetime.now().timestamp())}'
            }
            
        except Exception as e:
            self.logger.error(f"Error deleting account: {e}")
            return {'success': False, 'error': 'Failed to process account deletion'}
    
    def _get_subscription_info(self, user_id: str) -> Dict[str, Any]:
        """Get subscription information for user"""
        # In a real implementation, this would query the subscription database
        return {
            'tier': 'Free',
            'monthly_cost': 0,
            'api_calls_used': 245,
            'api_calls_limit': 1000,
            'features_enabled': ['Basic Trading', 'AI Analysis', 'Portfolio Tracking'],
            'upgrade_available': True
        }
    
    def _get_tier_features(self, tier: str) -> List[str]:
        """Get features for subscription tier"""
        features_map = {
            'Free': [
                'Basic AI Trading Insights',
                'Portfolio Tracking',
                'Stock Search & Analysis',
                '1,000 API Calls/Month',
                'Email Support'
            ],
            'Pro': [
                'Advanced AI Recommendations',
                'Real-time Market Data',
                'Level 2 Data Access',
                'Smart Order Routing',
                '10,000 API Calls/Month',
                'Priority Support'
            ],
            'Elite': [
                'All Pro Features',
                'Dark Pool Intelligence',
                'Advanced Options Flow',
                'Algorithm Builder',
                '50,000 API Calls/Month',
                'Phone Support'
            ],
            'Institutional': [
                'All Elite Features',
                'Direct Market Access',
                'Custom API Integration',
                'Team Management',
                'Unlimited API Calls',
                'Dedicated Account Manager'
            ]
        }
        return features_map.get(tier, [])

# Initialize the account settings manager
account_settings_manager = AccountSettingsManager()